### PR TITLE
Added static and dynamic validations for tempate parameters

### PR DESCRIPTION
**Pull request by anuragc617**

### CODE REVIEW

File: src/configure/clients/azure/armRestClient.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export class ArmRestClient {
         this.restClient = new RestClient(azureSession.credentials);
     }
 
-    public fetchArmData(endPointUri: string, httpMethod: string, body?: string) {
+    public fetchArmData(endPointUri: string, httpMethod: string, body?: any) {
```
Comment (bishal-pdMSFT) [inline src/configure/clients/azure/armRestClient.ts:13]: why this change?

File: src/configure/configure.ts [hunk 2]
```diff
@@ -446,7 +446,7 @@ class Orchestrator {
             repoAnalysisResult = await repoAnalysisHelper.getRepositoryAnalysis(this.inputs.sourceRepository, this.inputs.pipelineConfiguration.workingDirectory.split('\\').join('/'));
         }
 
-        extensionVariables.templateServiceEnabled = false;
+        extensionVariables.templateServiceEnabled = true;
```
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:449]: please turn it off

File: src/configure/configure.ts [hunk 3]
```diff
@@ -513,7 +513,7 @@ class Orchestrator {
         }
 
         if (extensionVariables.templateServiceEnabled) {
-            telemetryHelper.setTelemetry(TelemetryKeys.ChosenTemplate, this.inputs.pipelineConfiguration.templateInfo.label);
+            telemetryHelper.setTelemetry(TelemetryKeys.ChosenTemplate, this.inputs.pipelineConfiguration.templateInfo.templateDescription);
```
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:516]: this might cause an issue if we want to look at both current and historical data. the format is changing.
Should we rather log `<label>:<targetKind>` ?

File: src/configure/model/templateModels.ts [hunk 4]
```diff
@@ -17,10 +17,10 @@ export interface PipelineTemplate {
 
 export interface PipelineTemplateMetadata {
     templateId: string;
-    label: string;
+    templateDescription: string;
```
Comment (bishal-pdMSFT) [inline src/configure/model/templateModels.ts:20]: Okay I see you have renamed this. But while logging telemetry, I still think you should append target kind if it exists

Comment (bishal-pdMSFT) [inline src/configure/model/templateModels.ts:20]: just `description`

File: src/configure/templateInputHelper/InputControl.ts [hunk 5]
```diff
@@ -62,21 +73,31 @@ export class InputControl {
...
             var dependentInputs = this._getDataSourceInputs();
             if (this.controlType === ControlType.None || this.controlType === ControlType.InputBox) {
-                this.value = await this.dataSource.evaluateDataSources(dependentInputs, azureSession);
+                this.value = await this.dataSource.evaluateDataSources(dependentInputs, this.azureSession);
             }
             else if (this.controlType === ControlType.QuickPick) {
-                let selectedValue = await this.dataSource.evaluateDataSources(dependentInputs, azureSession)
-                    .then((listItems: Array<{ label: string, data: any, group?: string }>) => {
-                        return this.showQuickPick(this.getInputControlId(), listItems, { placeHolder: this.inputDescriptor.name });
+                let listItems: Array<QuickPickItemWithData> = await vscode.window.withProgress(
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:86]: Nice

Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:86]: Nit: Should the return type be QuickPickItemWithData or something more generic?

File: src/configure/templateInputHelper/InputControl.ts [hunk 6]
```diff
@@ -62,21 +73,31 @@ export class InputControl {
...
+                    { location: vscode.ProgressLocation.Notification, title: utils.format(Messages.fetchingInputMessage, this.inputDescriptor.name) },
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:87]: question: what other location is possible?

File: src/configure/templateInputHelper/InputControl.ts [hunk 7]
```diff
@@ -62,21 +73,31 @@ export class InputControl {
...
-    public async setInputControlValue(azureSession: AzureSession): Promise<any> {
+    public async setInputControlValue(): Promise<any> {
         if (!this.isVisible()) {
             return;
         }
         if (!!this.dataSource) {
             var dependentInputs = this._getDataSourceInputs();
             if (this.controlType === ControlType.None || this.controlType === ControlType.InputBox) {
-                this.value = await this.dataSource.evaluateDataSources(dependentInputs, azureSession);
+                this.value = await this.dataSource.evaluateDataSources(dependentInputs, this.azureSession);
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:83]: Won't we need to trigger dynamic/static validation for input box?

Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControl.ts:83]: ahh okay we are only getting the value here

File: src/configure/templateInputHelper/InputControl.ts [hunk 8]
```diff
@@ -95,10 +116,9 @@ export class InputControl {
             }
             else if (this.controlType === ControlType.InputBox) {
                 this.value = await this.showInputBox(this.getInputControlId(), {
+                    value: this.value,
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:119]: why is this change required?

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 9]
```diff
@@ -1,58 +1,62 @@
...
 import { VisibilityHelper } from "./utilities/VisibilityHelper";
 
 export class InputControlProvider {
     private _pipelineTemplate: ExtendedPipelineTemplate;
     private _inputControlsMap: Map<string, InputControl>;
+    private azureSession: AzureSession;
+    private _context: StringMap<any>;
 
-    constructor(pipelineTemplate: ExtendedPipelineTemplate, context: { [key: string]: any }) {
+    constructor(azureSession: AzureSession, pipelineTemplate: ExtendedPipelineTemplate, context: { [key: string]: any }) {
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:16]: use StringMap here as well?

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 10]
```diff
@@ -1,58 +1,62 @@
...
+        this._createControls();
     }
 
-    public async getAllPipelineTemplateInputs(azureSession: AzureSession, resourceNode?: GenericResource) {
+    public async getAllPipelineTemplateInputs() {
         let parameters: { [key: string]: any } = {};
         for (let inputControl of this._inputControlsMap.values()) {
-            if (inputControl.getPropertyValue('deployTarget') === "true" && !!resourceNode) {
-                inputControl.setValue(resourceNode.id);
+            if (inputControl.getPropertyValue('deployTarget') === "true" && !!this._context["resourceId"]) {
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:27]: using an index into a map can lead to very cryptic error messages like `element not found`. you can define a function like `getItem(key: string)` which will return a value or will throw a friendly error message

Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:27]: If I am not wrong, dependency on this property `deployTarget` is subject to change based on how we want to surface this information in template. right?

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 11]
```diff
@@ -81,6 +86,42 @@ export class InputControlProvider {
...
+                    var validationDataSource = DataSourceUtility.getDataSourceById(this._pipelineTemplate.dataSources, validation.dataSourceId);
+                    if (validationDataSource) {
+                        dataSourceToInputsMap.set(validationDataSource, []);
+
+                        let dynamicValidationRequiredInputIds = DataSourceUtility.getDependentInputIdList(validationDataSource.endpointUrlStem);
+                        dynamicValidationRequiredInputIds = dynamicValidationRequiredInputIds.concat(DataSourceUtility.getDependentInputIdList(validationDataSource.requestBody));
+                        dynamicValidationRequiredInputIds = dynamicValidationRequiredInputIds.concat(DataSourceUtility.getDependentInputIdList(validationDataSource.resultTemplate));
+                        dynamicValidationRequiredInputIds = dynamicValidationRequiredInputIds.concat(DataSourceUtility.getDependentInputIdList(validationDataSource.resultSelector));
+
+                        dynamicValidationRequiredInputIds = Array.from(new Set(dynamicValidationRequiredInputIds));
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:106]: why do we need this Array.from ?

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 12]
```diff
@@ -81,6 +86,42 @@ export class InputControlProvider {
...
+                        dynamicValidationRequiredInputIds.forEach((dynamicValidationRequiredInputId) => {
+                            var dependentInput = this._inputControlsMap.get(dynamicValidationRequiredInputId);
+                            if (dependentInput && dynamicValidationRequiredInputId !== inputDes.id) {
+                                dataSourceToInputsMap.get(validationDataSource).push(dependentInput);
+                            } else {
+                                //.logError("Dependent input {0} specified for input {1} is not present in pipeline template {2}".format(dynamicValidationRequiredInputId, inputDes.id, this._pipelineTemplate.id));
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:112]: we can log trace for this

File: src/configure/templateInputHelper/InputControlProvider.ts [hunk 13]
```diff
@@ -81,6 +86,42 @@ export class InputControlProvider {
...
+                            }
+                        });
+                    } else {
+                        //PipelineTemplateErrorLogger.logError("validation data source {0} specified for input {1} is not present in pipeline template {2}".format(validation.dataSourceId, inputDes.id, this._pipelineTemplate.id));
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:116]: we can log trace for this

Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControlProvider.ts:116]: actually we should log only if `dependentInput` is null.

Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControlProvider.ts:116]: it would be good to trace these errors, as they help in figuring out if there are some issues in our templates

File: src/configure/helper/mustacheHelper.ts [hunk 14]
```diff
@@ -95,6 +95,25 @@ export class MustacheHelper {
                     }
 
                     return "";
+                };
+            },
+            "equals": function () {
+                return function (text: string, render: any) {
```
Comment (kanika1894) [inline src/configure/helper/mustacheHelper.ts:101]: As we discussed, I have already implemented equals in my rightClickScenario PR

File: src/configure/configure.ts (hunk 3 above)
Comment (kanika1894) [reply src/configure/configure.ts:516]: This seems to be pretty old code as ``` pipelineConfiguration.templateInfo ``` doesn't exist anymore.

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 10 above)
Comment (kanika1894) [reply src/configure/templateInputHelper/InputControlProvider.ts:27]: Yes, this part is changed as now we'll be using the concept of override in properties. This was initially done to handle right click scenario.

File: src/configure/clients/github/TemplateServiceClient.ts [hunk 15]
```diff
@@ -20,7 +19,7 @@ export class TemplateServiceClient {
             body);
     }
 
-    public async getTemplateParameters(templateId: string): Promise<ExtendedPipelineTemplate> {
+    public async getTemplateParameters(templateId: string): Promise<any> {
```
Comment (kanika1894) [inline src/configure/clients/github/TemplateServiceClient.ts:22]: Ideally we should have a fix return type, right? so, any specific reason for this change?

File: src/configure/templateInputHelper/InputControl.ts [hunk 16]
```diff
@@ -123,6 +173,92 @@ export class InputControl {
...
+                    validationResult = validationResult.concat(result);
+                }
+            }
+        }
+        return validationResult;
+    }
+
+    private async evaluateDynamicValidation(validation: InputDynamicValidation, value: string): Promise<string> {
+        var requiredDataSource: DataSource = null;
+        this.validationDataSourceToInputsMap.forEach((descriptorsArray, dataSource) => { if (dataSource.id === validation.dataSourceId) { requiredDataSource = dataSource; } });
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:211]: add types for descriptorsArray and dataSource

Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:211]: please format the code

Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControl.ts:211]: in multiple lines

Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:211]: Shouldn't the order of arguments be reversed - (dataSource, descriptorsArray) as datasource is key here?

File: src/configure/templateInputHelper/InputControl.ts [hunk 17]
```diff
@@ -110,11 +130,41 @@ export class InputControl {
...
         if (!!properties) {
             value = properties[propertyName];
-            if( !!value && !!inputs ) { return MustacheHelper.render(value, { inputs: inputs }); }
+            if (!!value && !!inputs) { return MustacheHelper.render(value, { inputs: inputs }); }
         }
         return value;
     }
 
+    public setValidationDataSources(dataSourceToInputsMap: Map<DataSource, InputControl[]>) {
+        this.validationDataSourceToInputsMap = dataSourceToInputsMap;
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:139]: Do we really need this map? What purpose does it serve? Should we just a list of <DataSource> as dynamic validations for an input are already stored in `this.inputDescriptor.dynamicValidations`

File: src/configure/templateInputHelper/InputControl.ts [hunk 18]
```diff
@@ -110,11 +130,41 @@ export class InputControl {
...
+    }
+
+    public async triggerControlValueValidations(value: string): Promise<string> {
+        this._tempValue = value;
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/InputControl.ts:143]: temp is an ambiguous word. If I understand correctly it is the value which has been selected by user, has not yet been validated. If so, should we call it _valuePendingValidation  or something similar?

File: src/configure/templateInputHelper/utilities/DataSourceUtility.ts [hunk 19]
```diff
@@ -39,7 +40,7 @@ export class DataSourceUtility {
         var httpMethod = dataSource.httpMethod || "GET";
         var requestBody = !!dataSource.requestBody ? MustacheHelper.render(dataSource.requestBody, view) : null;
         let amrClient = new ArmRestClient(azureSession);
-        return amrClient.fetchArmData(armUri, httpMethod, requestBody)
+        return amrClient.fetchArmData(armUri, httpMethod, JSON.parse(requestBody))
```
Comment (bishal-pdMSFT) [inline src/configure/templateInputHelper/utilities/DataSourceUtility.ts:43]: why parse?

File: src/configure/clients/github/TemplateServiceClient.ts (hunk 15 above)
Comment (bishal-pdMSFT) [reply src/configure/clients/github/TemplateServiceClient.ts:22]: +1

Comment (anuragc617) [reply src/configure/clients/github/TemplateServiceClient.ts]: While testing with default inputs I changed it, forgot to change back.

File: src/configure/configure.ts (hunk 2 above)
Comment (anuragc617) [reply src/configure/configure.ts]: done

File: src/configure/configure.ts (hunk 3 above)
Comment (anuragc617) [reply src/configure/configure.ts]: yes it doesn't exist, this branch was not updated after your changes.

File: src/configure/templateInputHelper/InputControl.ts (hunk 18 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts]: yes, changed it.

File: src/configure/templateInputHelper/InputControl.ts (hunk 16 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts]: Done

Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts]: done

Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts]: foreach in typescript has reverse order for Map -> (value, key)

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 11 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControlProvider.ts:110]: because dynamicValidationRequiredInputIds is of Array type.

File: src/configure/templateInputHelper/utilities/DataSourceUtility.ts (hunk 19 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/utilities/DataSourceUtility.ts:43]: MustacheHelper.render returns string, and the RestClient that we use for making API calls does stringify of body. If JSON object is not passed then it will become invalid string enclosed in two double quotes after stringify.

File: src/configure/templateInputHelper/InputControl.ts (hunk 17 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts:139]: this.inputDescriptor.dynamicValidations contains only dynamicValidations source Id, this is map of datasource  and required inputControls to evaluate the dynamic validation dataSource  which is set by InputControlProvider.

File: src/configure/templateInputHelper/InputControl.ts (hunk 5 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts:86]: thanks

File: src/configure/helper/mustacheHelper.ts (hunk 14 above)
Comment (anuragc617) [reply src/configure/helper/mustacheHelper.ts]: Needed to evaluate some dynamic validation conditions, removed it from my PR.

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 10 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControlProvider.ts:31]: in typescript it return undefined.

File: src/configure/templateInputHelper/InputControl.ts (hunk 7 above)
Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControl.ts:83]: What if the value returned by datasource is invalid? 
We should still run validation and if validation fails then show the input to user. What do you think? Otherwise user will get an error and they will not even know hot fix it.

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 11 above)
Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControlProvider.ts:110]: But `dynamicValidationRequiredInputIds` is also an array. No?

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 13 above)
Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControlProvider.ts]: I don't see any trace statement here

File: src/configure/templateInputHelper/InputControl.ts (hunk 17 above)
Comment (bishal-pdMSFT) [reply src/configure/templateInputHelper/InputControl.ts:139]: Got it.

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 11 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControlProvider.ts:110]: here first a 'Set' is created to get only 'unique values' and then it is converted back to Array.

File: src/configure/templateInputHelper/InputControlProvider.ts (hunk 13 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControlProvider.ts]: you looked at older commit, trace statement is there.

File: src/configure/templateInputHelper/InputControl.ts (hunk 7 above)
Comment (anuragc617) [reply src/configure/templateInputHelper/InputControl.ts:83]: yes that's right, but user may not know the correct value for it. But he should have the option to fix the value. Added it.
